File: mcspring-api/mcspring-subcommands/src/main/java/in/kyle/mcspring/command/TabCommandFactory.java
package in.kyle.mcspring.command;

import org.bukkit.command.Command;
import org.bukkit.command.PluginCommand;
import org.bukkit.command.TabCompleter;
import org.bukkit.plugin.Plugin;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.util.Set;
import java.util.function.Consumer;

import in.kyle.mcspring.subcommands.tab.TabDiscovery;
import lombok.SneakyThrows;
import lombok.var;

@Primary
@Component
class TabCommandFactory extends SimpleCommandFactory {
    
    private final TabDiscovery tabDiscovery;
    
    public TabCommandFactory(SimpleMethodInjection injection,
                             Set<CommandResolver> commandResolvers,
                             Plugin plugin,
                             TabDiscovery tabDiscovery) {
        super(injection, commandResolvers, plugin);
        this.tabDiscovery = tabDiscovery;
    }
    
    @Override
    public Command makeCommand(Method method, Object object, String name) {
        var command = (PluginCommand) super.makeCommand(method, object, name);
        if (method.getParameterCount() == 1 && method.getParameters()[0].getType()
                .isAssignableFrom(in.kyle.mcspring.subcommands.PluginCommand.class)) {
            command.setTabCompleter(makeTabCompleter(method, object));
        }
        return command;
    }
    
    private Consumer<in.kyle.mcspring.subcommands.PluginCommand> methodToConsumer(Method method,
                                                                                  Object object) {
        return pluginCommand -> {
            method.setAccessible(true);
            invoke(method, object, pluginCommand);
        };
    }
    
    @SneakyThrows
    private Object invoke(Method method,
                          Object object,
                          in.kyle.mcspring.subcommands.PluginCommand command) {
        return method.invoke(object, command);
    }
    
    private TabCompleter makeTabCompleter(Method method, Object object) {
        return (sender, command, s, strings) -> {
            var consumer = methodToConsumer(method, object);
            return tabDiscovery.getCompletions(sender, String.join(" ", strings), consumer);
        };
    }
}

File: mcspring-api/mcspring-subcommands/src/main/java/in/kyle/mcspring/subcommands/tab/TabCommandFactory.kt
package `in`.kyle.mcspring.subcommands.tab

import `in`.kyle.mcspring.command.CommandResolver
import `in`.kyle.mcspring.command.SimpleCommandFactory
import `in`.kyle.mcspring.command.SimpleMethodInjection
import `in`.kyle.mcspring.subcommands.plugincommand.api.PluginCommand
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.plugin.Plugin
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.context.annotation.Primary
import org.springframework.stereotype.Component
import java.lang.reflect.Method
import org.bukkit.command.Command as BukkitCommand
import org.bukkit.command.PluginCommand as BukkitPluginCommand

@Primary
@Component
@ConditionalOnBean(Plugin::class)
internal class TabCommandFactory(
        injection: SimpleMethodInjection,
        commandResolvers: Set<CommandResolver>,
        val plugin: Plugin,
        private val tabDiscovery: TabDiscovery
) : SimpleCommandFactory(injection, commandResolvers, plugin) {

    override fun makeCommand(method: Method, obj: Any, name: String): BukkitCommand {
        val command = super.makeCommand(method, obj, name) as BukkitPluginCommand
        if (method.parameterCount == 1
                && method.parameters[0].type.isAssignableFrom(PluginCommand::class.java)) {
            command.tabCompleter = makeTabCompleter(method, obj)
        }
        return command
    }

    private fun methodToConsumer(method: Method, obj: Any): (PluginCommand) -> Unit {
        return {
            method.isAccessible = true
            method.invoke(obj, it)
        }
    }

    private fun makeTabCompleter(method: Method, obj: Any): TabCompleter {
        return TabCompleter { sender: CommandSender, _: BukkitCommand, _: String, strings: Array<String> ->
            val consumer = methodToConsumer(method, obj)
            tabDiscovery.getCompletions(sender, strings.joinToString(" "), consumer)
        }
    }
}
